### PR TITLE
GCS and S3 CI/CD templates

### PR DESCRIPTION
**Pull request by jwitz**
A rework of https://github.com/astronomer/docs/pull/2099 using our new CI/CD documentation.

### CODE REVIEW

File: astro/ci-cd-templates/gcs.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+      download_to_local('my-demo-bucket', 'dags', '/tmp/astro/dags')
+      
+      ## download astro cli binary and move to /tmp/astro
+      download_to_local('my-demo-bucket', 'cli_binary', '/tmp/astro')
+      untar('/tmp/astro/astro_cli.tar.gz')
+      ## deploy to astro
+      os.chdir('/tmp/astro')
+      run_command('echo y | ./astro dev init')
+      run_command('./astro deploy --dags')
+  ```
```
Comment (fritz-astronomer) [inline astro/ci-cd-templates/gcs.md:108]:
```python
import os
import tarfile
import subprocess
from pathlib import Path

from google.cloud import storage
from google.cloud.functions import Context

BUCKET = os.getenv("BUCKET", "my-demo-bucket")


def untar(filename: str, destination: str) -> None:
    with tarfile.open(filename) as file:
        file.extractall(destination)


def run_command(cmd: str) -> None:
    p = subprocess.Popen("set -x; " + cmd, shell=True)
    p.communicate()


def download_to_local(bucket_name: str, gcs_folder: str, local_dir: str = None) -> None:
    """Download the contents of a folder directory
    :param bucket_name: the name of the gcs bucket
    :param gcs_folder: the folder path in the gcs bucket
    :param local_dir: a relative or absolute directory path in the local file system
    """
    storage_client = storage.Client()
    source_bucket = storage_client.bucket(bucket_name)
    blobs = source_bucket.list_blobs(prefix=gcs_folder)
    for blob in blobs:
        path = Path(local_dir) / blob.name if local_dir else Path(blob.name)
        if not path.is_dir():
            path.parent.mkdir(parents=True, exist_ok=True)
            blob.download_to_filename(path.absolute())


def astro_deploy(event: dict, context: Context) -> None:
    """Triggered by a change to a Cloud Storage bucket.
    :param event: Event payload.
    :param context: Metadata for the event.
    """
    os.chdir('/tmp/astro')
    download_to_local(BUCKET, 'dags', './dags')
    download_to_local(BUCKET, 'cli_binary', '.')
    untar('./astro_cli.tar.gz', '.')
    run_command('echo y | ./astro dev init')
    run_command('./astro deploy --dags')
```

File: astro/ci-cd-templates/gcs.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      ## download dag files to temp local storage
+      download_to_local('my-demo-bucket', 'dags', '/tmp/astro/dags')
+      
+      ## download astro cli binary and move to /tmp/astro
+      download_to_local('my-demo-bucket', 'cli_binary', '/tmp/astro')
+      untar('/tmp/astro/astro_cli.tar.gz')
+      ## deploy to astro
+      os.chdir('/tmp/astro')
+      run_command('echo y | ./astro dev init')
+      run_command('./astro deploy --dags')
```
Comment (fritz-astronomer) [inline astro/ci-cd-templates/gcs.md:107]:
```suggestion
```python
import os
import tarfile
import subprocess
from pathlib import Path

from google.cloud import storage
from google.cloud.functions import Context

BUCKET = os.getenv("BUCKET", "my-demo-bucket")


def untar(filename: str, destination: str) -> None:
    with tarfile.open(filename) as file:
        file.extractall(destination)


def run_command(cmd: str) -> None:
    p = subprocess.Popen("set -x; " + cmd, shell=True)
    p.communicate()


def download_to_local(bucket_name: str, gcs_folder: str, local_dir: str = None) -> None:
    """Download the contents of a folder directory
    :param bucket_name: the name of the gcs bucket
    :param gcs_folder: the folder path in the gcs bucket
    :param local_dir: a relative or absolute directory path in the local file system
    """
    storage_client = storage.Client()
    source_bucket = storage_client.bucket(bucket_name)
    blobs = source_bucket.list_blobs(prefix=gcs_folder)
    for blob in blobs:
        path = Path(local_dir) / blob.name if local_dir else Path(blob.name)
        if not path.is_dir():
            path.parent.mkdir(parents=True, exist_ok=True)
            blob.download_to_filename(path.absolute())


def astro_deploy(event: dict, context: Context) -> None:
    """Triggered by a change to a Cloud Storage bucket.
    :param event: Event payload.
    :param context: Metadata for the event.
    """
    os.chdir('/tmp/astro')
    download_to_local(BUCKET, 'dags', './dags')
    download_to_local(BUCKET, 'cli_binary', '.')
    untar('./astro_cli.tar.gz', '.')
    run_command('echo y | ./astro dev init')
    run_command('./astro deploy --dags')
```

File: astro/ci-cd-templates/gcs.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+- A [Deployment API key ID and secret](api-keys.md).
+- An [Astro project](create-first-dag.md) containing your project configurations.
+
+### DAG-based deploy
+
+This CI/CD template can be used to deploy DAGs from a single GCS bucket to a single Astro Deployment. When you create or modify a DAG in the GCS bucket, a Lambda function triggers to deploy your DAGs to Astro.
+
+:::info
+
+To deploy any non-DAG code changes to Astro, you need to trigger a standard image-only deploy with your Astro project. When you do this, your Astro project must include the latest version of your DAGs from your GCS bucket. If your Astro project `dags` folder isn't up to date with your GCS DAGs bucket when you trigger this deploy, you will revert your DAGs back to the version hosted in your Astro project.
```
[the remark this answers is not in the document]
Comment (manmeetkaur) [reply astro/ci-cd-templates/gcs.md:21]: @jwitz so there are two parts to CI/CD when deploying to Astro with DAG-only Enabled from Cloud Storage:

1. Folks will still use CI/CD to publish the DAGs to cloud storage (gcs or s3) 
2. Lambda/GCF will then auto-trigger based on the changes to the bucket

When there are changes to non-DAG resources, like requirements.txt, custom operators or other helpers, then the CI/CD should do image deploy automatically. For this we have can create a separate template which checks changes in `requirements.txt` or `include` and do image deploy ELSE just push the DAGs to the bucket which will trigger DAG-only deploy. I can create a template for that if you would like.

[the remark this answers is not in the document]
Comment (manmeetkaur) [reply astro/ci-cd-templates/gcs.md:21]: agreed @jwitz . We can move forward. I will just update GCS code per the suggestions from @fritz-astronomer and push it.

File: astro/ci-cd-templates/aws-s3.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+## Prerequisites
+
+- An AWS S3 bucket
+- An [Astro Deployment](create-deployment.md) with [DAG-only deploys enabled](deploy-code.md#enable-dag-only-deploys-on-a-deployment).
+- A [Deployment API key ID and secret](api-keys.md).
+- An [Astro project](create-first-dag.md) containing your project configurations.
+
+### DAG-based deploy
+
+This CI/CD template can be used to deploy DAGs from a single S3 bucket to a single Astro Deployment. When you create or modify a DAG in the S3 bucket, a Lambda function triggers to deploy your DAGs to Astro.
```
Comment (lzdanski) [inline astro/ci-cd-templates/aws-s3.md:19]:
```suggestion
This CI/CD template can be used to deploy DAGs from a single S3 bucket to a single Astro Deployment. Astro Deployment. When you create or modify a DAG in the S3 bucket, a Lambda function triggers your project to deploy DAGs to Astro.
```

File: astro/ci-cd-templates/aws-s3.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+:::info
+
+To deploy any non-DAG code changes to Astro, you need to trigger a standard image-only deploy with your Astro project. When you do this, your Astro project must include the latest version of your DAGs from your S3 bucket. If your Astro project `dags` folder isn't up to date with your S3 DAGs bucket when you trigger this deploy, you will revert your DAGs back to the version hosted in your Astro project.
+
+:::
+
+1. Download the latest Astro CLI binary from [GitHub releases](https://github.com/astronomer/astro-cli/releases), then rename the file to, `astro_cli.tar.gz`. For example, to use Astro CLI version 1.13.0 in your template, you would download `astro_1.13.0_linux_amd64.tar.gz` and rename it to `astro_cli.tar.gz`.
```
Comment (lzdanski) [inline astro/ci-cd-templates/aws-s3.md:27]:
```suggestion
1. Download the latest Astro CLI binary from [GitHub releases](https://github.com/astronomer/astro-cli/releases), then rename the file to, `astro_cli.tar.gz`. For example, to use Astro CLI version 1.13.0 in your template, download `astro_1.13.0_linux_amd64.tar.gz` and rename it to `astro_cli.tar.gz`.
```

**Review by lzdanski — CHANGES_REQUESTED** [review]
Some passive/active voice nits!

**Review by manmeetkaur — APPROVED** [review]

File: astro/ci-cd-templates/aws-s3.md (hunk 4 above)
Comment (manmeetkaur) [reply astro/ci-cd-templates/aws-s3.md]: changes done 👌

Comment (manmeetkaur) [reply astro/ci-cd-templates/aws-s3.md]: changes done.

File: astro/ci-cd-templates/gcs.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    
+        ## download astro cli binary and move to /tmp/astro
+        download_to_local(BUCKET, 'cli_binary', base_dir)
+
+        ## deploy to astro
+        os.chdir(base_dir)
+        untar('./astro_cli.tar.gz', '.')
+        run_command('echo y | ./astro dev init')
+        run_command(f'./astro deploy --dags')
+  ```
```
Comment (jwitz) [inline astro/ci-cd-templates/gcs.md:118]:
```suggestion
    ```
```

**Pull request merged**